src/request/providers/node.ts

```diff
@@ -0,0 +1,558 @@
+							data.buffer.push(chunk);
+						}
+
+						data.size += typeof chunk === 'string' ?
+							Buffer.byteLength(chunk, options.streamEncoding) :
+							chunk.length;
+
+						response.emit({
+							type: 'progress',
+							response,
+							totalBytesDownloaded: data.size
+						});
+					});
+
+					message.once('end', () => {
+						timeoutHandle && timeoutHandle.destroy();
+
+						let dataAsBuffer = (options.streamEncoding ? new Buffer(data.buffer.join(''), 'utf8') : Buffer.concat(data.buffer, data.size));
+
+						const handleEncoding = function () {
+							/*
+							 Content encoding is ordered by the order in which they were applied to the
+							 content, so do undo the encoding we have to start at the end and work backwards.
+							 */
+							if (contentEncodings.length) {
+								const encoding = contentEncodings.pop()!.trim();
+
+								if (encoding === '' || encoding === 'identity') {
+									// do nothing, response stream is as-is
+									handleEncoding();
+								}
+								else if (encoding === 'gzip') {
+									zlib.gunzip(dataAsBuffer, function (err: Error, result: Buffer) {
+										if (err) {
+											reject(err);
+										}
+
+										dataAsBuffer = result;
+										handleEncoding();
+									});
+								}
+								else if (encoding === 'deflate') {
+									zlib.inflate(dataAsBuffer, function (err: Error, result: Buffer) {
+										if (err) {
+											reject(err);
+										}
+
+										dataAsBuffer = result;
+										handleEncoding();
+									});
+								}
+								else {
+									reject(new Error('Unsupported content encoding, ' + encoding));
+								}
+							}
+							else {
+								data.data = String(dataAsBuffer);
+
+								response.emit({
+									type: 'end',
+									response
+								});
+
+								resolve(message);
+							}
+						};
+
+						handleEncoding();
+					});
+				});
+			}, () => {
+				request.abort();
+			});
+
+			const data: RequestData = {
+				task,
+				buffer: [],
+				data: '',
+				size: 0,
+				used: false,
+				url: url,
+				requestOptions: options,
+				nativeResponse: message
+			};
+
+			dataMap.set(response, data);
+
+			resolve(response);
+		});
+
+		request.once('error', reject);
+
+		if (options.body) {
+			if (options.body instanceof Buffer) {
+				request.end(options.body.toString());
+			}
+			else {
+				request.end(options.body.toString());
+			}
+		}
+		else {
+			request.end();
+		}
+
+		if (options.timeout > 0 && options.timeout !== Infinity) {
+			timeoutHandle = createTimer(() => {
+				timeoutReject && timeoutReject(new TimeoutError('The request timed out'));
+			}, options.timeout);
+		}
+	}, () => {
+		request.abort();
+	}).catch(function (error: Error): any {
+		const parsedUrl = urlUtil.parse(url);
+
+		if (parsedUrl.auth) {
+			parsedUrl.auth = '(redacted)';
+		}
+
+		const sanitizedUrl = urlUtil.format(parsedUrl);
+
+		error.message = '[' + requestOptions.method + ' ' + sanitizedUrl + '] ' + error.message;
```

Comment:
I'm really not sure what the best practice is here. At this point, it might be best to just leave this and if we hit this in reality (rather than trying to cover every potential case), we can adjust.